File: crates/benches/fixtures/impact_analyzer.go
package impact

import (
	"sort"
)

// FileNode represents a file and the paths it imports.
type FileNode struct {
	Path    string
	Imports []string
	Symbols []string
}

// ImpactReport describes the blast radius of a changed file.
type ImpactReport struct {
	ChangedFile         string
	DirectDependents    []string
	TransitiveDependents []string
}

// TotalAffected returns the count of unique affected files.
func (r *ImpactReport) TotalAffected() int {
	seen := make(map[string]bool)
	for _, p := range append(r.DirectDependents, r.TransitiveDependents...) { seen[p] = true }
	return len(seen)
}

// Analyzer builds and queries the reverse dependency graph.
type Analyzer struct {
	nodes   map[string]*FileNode
	reverse map[string][]string
}

// New creates a new Analyzer.
func New() *Analyzer {
	return &Analyzer{nodes: make(map[string]*FileNode), reverse: make(map[string][]string)}
}

// Register adds a file node to the graph.
func (a *Analyzer) Register(node *FileNode) {
	a.nodes[node.Path] = node
	for _, imp := range node.Imports {
		a.reverse[imp] = append(a.reverse[imp], node.Path)
	}
}

// ComputeImpact returns the full impact report for a changed file.
func (a *Analyzer) ComputeImpact(path string) *ImpactReport {
	direct := a.reverse[path]
	sort.Strings(direct)
	visited := make(map[string]bool)
	for _, d := range direct { visited[d] = true }
	visited[path] = true
	queue := append([]string{}, direct...)
	var transitive []string
	for len(queue) > 0 {
		cur := queue[0]; queue = queue[1:]
		for _, dep := range a.reverse[cur] {
			if !visited[dep] {
				visited[dep] = true
				transitive = append(transitive, dep)
				queue = append(queue, dep)
			}
		}
	}
	sort.Strings(transitive)
	return &ImpactReport{ChangedFile: path, DirectDependents: direct, TransitiveDependents: transitive}
}

// MostImpactful returns the top N files sorted by direct dependent count.
func (a *Analyzer) MostImpactful(n int) []string {
	type item struct{ path string; count int }
	items := make([]item, 0, len(a.nodes))
	for p := range a.nodes { items = append(items, item{path: p, count: len(a.reverse[p])}) }
	sort.Slice(items, func(i, j int) bool { return items[i].count > items[j].count })
	if n > len(items) { n = len(items) }
	result := make([]string, n)
	for i := 0; i < n; i++ { result[i] = items[i].path }
	return result
}

File: crates/benches/fixtures/impact_analyzer.py
from __future__ import annotations

from collections import defaultdict, deque
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional


@dataclass
class FileNode:
    path: str
    symbols: list[str] = field(default_factory=list)
    imports: list[str] = field(default_factory=list)


@dataclass
class ImpactReport:
    changed_file: str
    direct_dependents: list[str]
    transitive_dependents: list[str]
    total_affected: int

    @property
    def all_affected(self) -> list[str]:
        seen: set[str] = set()
        result = []
        for path in self.direct_dependents + self.transitive_dependents:
            if path not in seen:
                seen.add(path)
                result.append(path)
        return result


class ImpactAnalyzer:
    def __init__(self) -> None:
        self._nodes: dict[str, FileNode] = {}
        self._reverse_deps: dict[str, set[str]] = defaultdict(set)

    def register(self, node: FileNode) -> None:
        self._nodes[node.path] = node
        for imp in node.imports:
            self._reverse_deps[imp].add(node.path)

    def compute_impact(self, changed_path: str) -> ImpactReport:
        direct = list(self._reverse_deps.get(changed_path, set()))
        transitive: list[str] = []
        visited: set[str] = set(direct) | {changed_path}
        queue = deque(direct)
        while queue:
            current = queue.popleft()
            for dep in self._reverse_deps.get(current, set()):
                if dep not in visited:
                    visited.add(dep)
                    transitive.append(dep)
                    queue.append(dep)
        return ImpactReport(
            changed_file=changed_path,
            direct_dependents=direct,
            transitive_dependents=transitive,
            total_affected=len(direct) + len(transitive),
        )

    def batch_impact(self, changed_paths: list[str]) -> dict[str, ImpactReport]:
        return {p: self.compute_impact(p) for p in changed_paths}

    def most_impactful(self, top_n: int = 10) -> list[tuple[str, int]]:
        scores = [(p, len(self._reverse_deps.get(p, set()))) for p in self._nodes]
        scores.sort(key=lambda x: x[1], reverse=True)
        return scores[:top_n]
